parsers/haskell.c

```diff
@@ -0,0 +1,355 @@
+					if (! (c == ' ' || c == '\t')) {
+						return c;
+					}
+				}
+			} while (c != EOF && c != '|');
+			st = Find_Constr;
+		}
+	}
+	return '\n';
+}
+
+static void findHaskellTags (int is_literate)
+{
+	vString *name = vStringNew ();
+	char token[1001], arg[1001];
+	int c;
+	int in_tex_lit_code = 0;
+	c = get_next_char();
+
+	while (c != EOF)
+	{
+		if (c == '\n') {
+			c = get_next_char();
+			continue;
+		}
+
+		if (isspace(c)) {
+			skip_rest_of_line();
+			c = get_next_char();
+			continue;
+		}
+		if (is_literate && !in_tex_lit_code) {
+			if (c == '>') {
+				c = getcFromInputFile();
+				if (c == ' ') {
+					c = get_next_char();
+					if (!isident(c)) {
+						skip_rest_of_line();
+						c = get_next_char();
+						continue;
+					}
+				} else {
+					skip_rest_of_line();
+					c = get_next_char();
+					continue;
+				}
+			} else if (c == '\\') {
+				int n = get_line(token);
+				if (strncmp(token, "begin{code}", 11) == 0) {
+					in_tex_lit_code = 1;
+					c = get_next_char();
+					continue;
+				} else {
+					if (n > 0 && token[n-1] != '\n')
+						skip_rest_of_line();
+					else
+						c = get_next_char();
+				}
+				continue;
+			} else {
+				skip_rest_of_line();
+				c = get_next_char();
+				continue;
+			}
+		}
+		if (is_literate && in_tex_lit_code && c == '\\') {
+			if (strncmp(token, "end{code}", 9) == 0) {
+				in_tex_lit_code = 0;
+				c = get_next_char();
+				continue;
+			}
+		}
+		token[0] = c;
+		if (!isident(c)) {
+			skip_rest_of_line();
+			c = get_next_char();
+			continue;
+		}
+		if (!get_token(token, 1)) {
+			c = get_next_char();
+			continue;
+		}
+		do {
+			if ((c = getcFromInputFile()) == EOF)
+				return;
+		} while (c == ' ' || c == '\t');
+		arg[0] = c;
+		get_token(arg, 1);
+		if (strcmp(token, "data") == 0 || strcmp(token, "newtype") == 0) {
+			add_tag(arg, K_TYPE, name);
+			c = inside_datatype(name);
+			continue;
+		}
+		if (strcmp(token, "type") == 0)
+			add_tag(arg, K_TYPE, name);
+		else if (strcmp(token, "module") == 0)
+			add_tag(arg, K_MODULE, name);
+		else if (strcmp(token, "instance") == 0 ||
+				 strcmp(token, "foreign") == 0 ||
+				 strcmp(token, "import") == 0)
+			;
+		else {
+			if (arg[0] != ':')
+				add_tag(token, K_FUNCTION, name);
+		}
+		skip_rest_of_line();
+		c = get_next_char();
+	}
+	vStringDelete(name);
+}
+
+static void findNormalHaskellTags (void)
+{
+	findHaskellTags (0);
+}
+
+static void findLiterateHaskellTags (void)
+{
+	findHaskellTags (1);
+}
+
+extern parserDefinition* HaskellParser (void)
+{
+	static const char *const extensions [] = { "hs", NULL };
+	parserDefinition* def  = parserNew ("Haskell");
+
+	def->kindTable  = HaskellKinds;
+	def->kindCount  = ARRAY_SIZE(HaskellKinds);
+	def->extensions = extensions;
+	def->parser     = findNormalHaskellTags;
+	return def;
+}
+
+extern parserDefinition* LiterateHaskellParser (void)
+{
+	static const char *const extensions [] = { "lhs", NULL };
+	parserDefinition* def = parserNew ("Literate Haskell");
```

Comment:
Let's avoid to use whitespace in a parser name.

Reply:
Done. By the way, this variant of the parser is unused right now (and we don't use it in Geany either, just the ordinary HaskellParser).